packages/client/src/devices/devices.ts

```diff
@@ -12,22 +12,47 @@ import {
 } from 'rxjs';
 import { getLogger } from '../logger';
 
-const getDevices = (constraints?: MediaStreamConstraints) => {
+/**
+ * Returns an Observable that emits the list of available devices
+ * that meet the given constraints.
+ *
+ * @param constraints the constraints to use when requesting the devices.
+ * @param kind the kind of devices to request.
+ */
+const getDevices = (
+  constraints: MediaStreamConstraints,
+  kind: MediaDeviceKind,
+) => {
   return new Observable<MediaDeviceInfo[]>((subscriber) => {
-    navigator.mediaDevices
-      .getUserMedia(constraints)
-      .then((media) => {
-        // in Firefox, devices can be enumerated after userMedia is requested
-        // and permissions granted. Otherwise, device labels are empty
-        navigator.mediaDevices.enumerateDevices().then((devices) => {
-          subscriber.next(devices);
-          // If we stop the tracks before enumerateDevices -> the labels won't show up in Firefox
-          disposeOfMediaStream(media);
-          subscriber.complete();
-        });
+    const enumerate = async () => {
+      let devices = await navigator.mediaDevices.enumerateDevices();
+      // some browsers report empty device labels (Firefox).
+      // in that case, we need to request permissions (via getUserMedia)
+      // to be able to get the device labels
+      const needsGetUserMedia = devices.some(
+        (device) => device.kind === kind && device.label === '',
+      );
+      if (needsGetUserMedia) {
+        let mediaStream: MediaStream | undefined;
+        try {
+          mediaStream = await navigator.mediaDevices.getUserMedia(constraints);
+          devices = await navigator.mediaDevices.enumerateDevices();
```

Comment:
We will already have devices info(we call enumerateDevices) as per line 28, isn't it?

Reply:
Yes, we do, but some browsers require "getUserMedia" to be called before we can retrieve the device's label.

That is why we first enumerate the devices naively, and we check if their labels are there. If not, we retrieve a temporary MediaStream (with getUserMedia) and we run the device enumeration again.

Reply:
Ohh, about the browser thing that was clear to me, about calling the `device enumeration again` was unclear. But now it is. Thanks @oliverlaz 😄